website/docs/guides/markdown-features/markdown-features-code-blocks.mdx

```diff
@@ -272,10 +272,104 @@ Prefer highlighting with comments where you can. By inlining highlight in the co
   ```
 ````
 
-In the future, we may extend the magic comment system and let you define custom directives and their functionalities. The magic comments would only be parsed if a highlight metastring is not present.
+Below, we will introduce how the magic comment system can be extended to define custom directives and their functionalities. The magic comments would only be parsed if a highlight metastring is not present.
 
 :::
 
+### Custom magic comments {#custom-magic-comments}
+
+`// highlight-next-line` and `// highlight-start` etc. are called "magic comments", because they will be parsed and removed, and their purposes are to add metadata to the next line, or the section that the pair of start- and end-comments enclose.
+
+You can declare custom magic comments through theme config. For example, you can register another magic comment that adds a `code-block-error-line` class name:
+
+<Tabs>
+<TabItem value="docusaurus.config.js">
+
+```js
+module.exports = {
+  themeConfig: {
+    prism: {
+      magicComments: [
+        // Remember to extend the default highlight class name as well!
+        {
+          className: 'theme-code-block-highlighted-line',
+          line: 'highlight-next-line',
+          block: {start: 'highlight-start', end: 'highlight-end'},
+        },
+        // highlight-start
+        {
+          className: 'code-block-error-line',
+          line: 'This will error',
+        },
+        // highlight-end
+      ],
+    },
+  },
+};
+```
+
+</TabItem>
+<TabItem value="src/css/custom.css">
+
+```css
+.code-block-error-line {
+  background-color: #ff000020;
+  display: block;
+  margin: 0 calc(-1 * var(--ifm-pre-padding));
+  padding: 0 var(--ifm-pre-padding);
+  border-left: 3px solid #ff000080;
+}
+```
+
+</TabItem>
+<TabItem value="myDoc.md">
+
+````md
+In TypeScript, types help prevent runtime errors.
+
+```ts
+function greet(name: string) {
+  // This will error
+  console.log(name.toUpper());
+  // .toUpper doesn't exist on string
+}
+```
+````
+
+</TabItem>
+
+</Tabs>
+
+````mdx-code-block
+<BrowserWindow>
+
+In TypeScript, types help prevent runtime errors.
+
+```ts
+function greet(name: string) {
+  // This will error
+  console.log(name.toUpper());
+  // .toUpper doesn't exist on string
+}
+```
+
+</BrowserWindow>
+````
```

Comment:
nice way to introduce the feature 👍 

Looks good enough to merge, but wonder if this wouldn't be better to:
- move example above config
- show simpler JS code block without TS text above

Reply:
> move example above config

This is consistent with all other demos.

> show simpler JS code block without TS text above

Not sure what you mean...?

Reply:
> This is consistent with all other demos.

I don't know, the feeling I have is that when clicking on tabs, I don't know if the demo is part of the tabs or not, and it's weird to see it moving up/down. Probably a tabs container design issue, but I think it would be mitigated by showing first what magic comments can enable you to build.

> Not sure what you mean...?

I mean not all users understand TypeScript, but most will understand simple JavaScript

There are ways to simplify this example, like:

```js
const username = null;

// show error
username.toUpperCase()
```

When I first saw the example, my first thought was: is the sentence "In TypeScript, types help prevent runtime errors." saying something I should take care of?

Reply:
> I don't know if the demo is part of the tabs or not,

Ah, I got you. We should provide an option in tabs to make it bordered.

> ```ts
> const username = null;
> 
> // show error
> username.toUpperCase()
> ```

Makes sense. This will be on my list of TODOs during the next docs batch update